src/data/tree.lean

```diff
@@ -81,4 +81,64 @@ def map {β} (f : α → β) : tree α → tree β
 | nil := nil
 | (node a l r) := node (f a) (map l) (map r)
 
+/-- A non-nil `unit_tree`; useful when we want an arbitrary value other than `nil` -/
+abbreviation non_nil [inhabited α] : tree α := node default nil nil
+
+@[simp] lemma non_nil_ne [inhabited α] : non_nil ≠ (@nil α) := by trivial
+
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp] def nodes : tree α → ℕ
+| nil := 0
+| (node _ a b) := a.nodes + b.nodes + 1
+
+/-- The number of leaves of a binary tree -/
+@[simp] def leaves : tree α → ℕ
```

Comment:
Honestly I wouldn't expect a function called `tree.leaves` to count the **number** of leaves. Can you think of a better name?

Reply:
How does `num_leaves` sound?